### PR TITLE
Add assertion hasSameHashCodeAs

### LINKED ISSUE #995: Feature Request: hasSameHashCodeAs
#### Summary

Provide a (shortcut) way to assert that the hashCode of two objects are equal.

#### Example

An alternative way to write
```java
        assertThat("foo".hashCode()).isEqualTo("foo".hashCode());
```

could be
```java
        assertThat("foo").hasSameHashCodeAs("foo");
```

While this example uses `String` it should work with any type of object, as every `Object` has a `hashCode`.

#### Java 8 specific ?

NO

### PR DESCRIPTION
**Pull request by Riggs333**
#### Check List:
* Fixes #995 
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

**Comment by joel-costigliola** [comment]
Thanks, I will review your PR shortly !

File: src/main/java/org/assertj/core/internal/Objects.java [hunk 1]
```diff
@@ -802,6 +804,37 @@ public boolean areEqualToComparingOnlyGivenFields(Object actual, Object other,
...
+    if (other == null) {
+      throw failures.failure(info, otherShouldNotBeNull());
+    }
+
+    if (actual.hashCode() != other.hashCode()) {
+      throw failures.failure(info, shouldHaveSameHashCode(actual, other));
+    }
+  }
+
+  private BasicErrorMessageFactory otherShouldNotBeNull() {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Objects.java]: put that in a `org.assertj.core.error.OtherShouldNotBeNull` similar to `ShouldNotBeNull`.

File: src/main/java/org/assertj/core/internal/Objects.java [hunk 2]
```diff
@@ -802,6 +804,37 @@ public boolean areEqualToComparingOnlyGivenFields(Object actual, Object other,
...
+    return new BasicErrorMessageFactory("%nExpecting other not to be null");
+  }
+
+  private <A> BasicErrorMessageFactory shouldHaveSameHashCode(A actual, Object other) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Objects.java]: Extract to a `org.assertj.core.error.ShouldHaveSameHashCode` class.

File: src/main/java/org/assertj/core/internal/Objects.java [hunk 3]
```diff
@@ -802,6 +804,37 @@ public boolean areEqualToComparingOnlyGivenFields(Object actual, Object other,
...
+    return new BasicErrorMessageFactory("%nExpecting%n  <%s>%nto have the same hashCode as <%s>", actual, other);
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Objects.java]: change `hashCode as <%s>`  to `hashCode as%n  <%s>`

File: src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  protected ObjectAssert<Jedi> invoke_api_method() {
+    return assertions.hasSameHashCodeAs(OTHER_JEDI);
+  }
+
+  @Override
+  protected void verify_internal_effects() {
+    verify(objects).assertHasSameHashCodeAs(getInfo(assertions), getActual(assertions), OTHER_JEDI);
+  }
+
+  @Test
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java]: move these tests in a `org.assertj.core.internal.objects.Objects_assertHasSameHashCodeAs_Test`, use `Objects_assertHasToString_Test` as an example (don't hardcode error messages but verify the proper error message factory methods are called).

File: src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  public void should_fail_if_other_is_null() {
+    thrown.expectAssertionError(format("%nExpecting other not to be null"));
+    Jedi jedi = new Jedi("Yoda", "Blue");
+
+    assertThat(jedi).hasSameHashCodeAs(null);
+  }
+
+  @Test
+  public void should_succeed_if_hashCode_is_equal() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java]: `should_succeed_if_hashCode_are_equal`

File: src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    return assertions.hasSameHashCodeAs(OTHER_JEDI);
+  }
+
+  @Override
+  protected void verify_internal_effects() {
+    verify(objects).assertHasSameHashCodeAs(getInfo(assertions), getActual(assertions), OTHER_JEDI);
+  }
+
+  @Test
+  public void should_fail_if_hashCode_is_not_equal() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/object/ObjectAssert_hasSameHashCodeAs_Test.java]: `should_fail_if_hashCodes_differ`

**Comment by Riggs333** [comment]
@joel-costigliola Thanks for this detailed feedback.
I'm very sorry for my late response. Actually I couldn't find the time to work on this issue. Unfortunately I won't find some time in the near future due to private concerns.

But maybe @matthiasbalke or @RudiKlassen have some time? We worked together on the implementation.

**Comment by joel-costigliola** [comment]
hey @Riggs333 it's totally not to have the time, we are contributing on our spare time, hope you sort out your private concerns, that's more important.
If @matthiasbalke or @RudiKlassen have the time to do it, great but no problem if they don't, I'll wait a bit for their answers.

**Comment by matthiasbalke** [comment]
hey @joel-costigliola, I should no have write permissions for the forked repo, so I'll try to update the PR as soon as possible within the next days / next week.

**Comment by joel-costigliola** [comment]
Thanks for the work guys, I only refactored the code a bit.

**Pull request closed without merging**